src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,341 @@
+          break;
+        }
+
+        // if model check passed:
+        extensions.push_back(p);
+
+        if (annotate_iso_peak_count)
+        {
+          // with "+ 1" the monoisotopic peak is counted as well
+          // current peak is the monoisotopic peak
+          iso_peak_count[current_peak] = i + 1;
+        }
+      } // cluster has been formed
+
+      if (has_min_isopeaks)
+      {
+        clusters.push_back(extensions);
+        charges_of_extensions.push_back(q);
+      }
+    } // all charges tested, clusters complete
+    // if current_peak is possible monoisotopic peak, pick the best of its clusters and annotate peaks with a feature number
+    if (!clusters.empty())
+    {
+      // pick cluster with largest size and highest charge (since all have the same monoisotopic peak)
+      UInt best_idx = 0;
+      if (clusters.size() > 1)
+      {
+        Size largest_size = 0;
+        Int best_cluster_charge = min_charge;
+        for (UInt i = 0; i != clusters.size(); ++i)
+        {
+          if ((clusters[i].size() > largest_size) || ((clusters[i].size() == largest_size) && (charges_of_extensions[i] > best_cluster_charge)))
+          {
+            largest_size = clusters[i].size();
+            best_cluster_charge = charges_of_extensions[i];
+            best_idx = i;
+          }
+        }
+      }
+
+      // save result
+      mono_isotopic_peak_charge[current_peak] = charges_of_extensions[best_idx];
+      for (UInt i = 0; i != clusters[best_idx].size(); ++i)
+      {
+        features[clusters[best_idx][i]] = feature_number;
+
+        if (add_up_intensity)
+        {
+          mono_iso_peak_intensity[current_peak] += old_spectrum[clusters[best_idx][i]].getIntensity();
+        }
+      }
+      ++feature_number;
+    }
+  }
+
+  // apply changes, i.e. select the indices which should survive
+  std::vector<Size> select_idx;
+
+  for (size_t i = 0; i != spec.size(); ++i)
+  {
+
+    Size z = mono_isotopic_peak_charge[i];
+    if (annotate_charge)
+    {
+      spec.getIntegerDataArrays()[charge_index].push_back((int)z);
+    }
+    if (annotate_iso_peak_count)
+    {
+      spec.getIntegerDataArrays()[iso_peak_count_index].push_back((int)iso_peak_count[i]);
+    }
+
+    if (!keep_only_deisotoped)
+    { // keep all unassigned peaks
+      if (features[i] < 0) // this peak is not part of a cluster
+      {
+        select_idx.push_back(i);
+        continue;
+      }
+    }
+
+    if (z != 0) // this is a monoisotopic peak
+    {
+      if (add_up_intensity)
+      {
+        spec[i].setIntensity(mono_iso_peak_intensity[i]);
+      }
+
+      // convert mono-isotopic peak with charge assigned by deisotoping
+      if (make_single_charged)
+      {
+        spec[i].setMZ(spec[i].getMZ() * z - (z - 1) * Constants::PROTON_MASS_U);
+      }
+      select_idx.push_back(i);
+    }
+  }
+
+  // properly subsets all datapoints (incl. dataArrays)
+  spec.select(select_idx);
+  spec.sortByPosition();
```

Comment:
reasonably sure you dont need to sort, since input is sorted and `select` will not change order.

Reply:
if make_single_charged == true sorting should be necessary as the peaks' mz values change. I added an if clause